### PR TITLE
Presenters

### PR DESCRIPTION
**Pull request by t-harps**

### CODE REVIEW

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -17,26 +17,16 @@ def humanize_date_and_time(date)
...
     output << "Materialize.toast('#{j alert}', 3000, 'red');" if alert
-
     output << "Materialize.toast('#{j notice}', 3000);" if notice
-
-    flash['errors'].each do |error|
-      output << "Materialize.toast('#{j error}', 5000, 'red');"
-    end if flash[:errors]
-
-    flash.keys.each { |flash_type| flash.send('discard', flash_type) }
+    flash["errors"].each { |error| output << "Materialize.toast('#{j error}', 5000, 'red');" } if flash[:errors]
```
Comment (dmitrytrager) [inline app/helpers/application_helper.rb]: Looks like too complicated for oneliner iterator
Maybe something like
```ruby
if flash[:errors]
  output = flash[:errors].reduce(output) do |output, error|
    output << "Materialize.toast('#{j error}', 5000, 'red');"
    output
  end
end
```

File: app/helpers/approvals_helper.rb [hunk 2]
```diff
@@ -1,50 +1,24 @@
...
-      { placeholder: 'App name', class: 'validate devs_typeahead', required: true }
-    elsif type == 'User'
-      { placeholder: 'email@email.com', class: 'validate', required: true }
-    end
-  end
-
   def approvals_approvable_name(approvable)
     if approvable.respond_to? :username
-      approvable.username.present? ? approvable.username : approvable.email.split('@').first
+      approvable.username.present? ? approvable.username : approvable.email.split("@").first
```
Comment (dmitrytrager) [inline app/helpers/approvals_helper.rb]: extracting name from email could be moved into separate method

File: app/helpers/permissions_helper.rb [hunk 3]
```diff
@@ -21,7 +21,7 @@ def permissions_control_class(permissionable)
   end
 
   def permissions_switch_class(permissionable)
-    permissionable.class == Permission ? 'permission-switch' : 'master'
+    permissionable.class == Permission ? "permission-switch" : "master"
```
Comment (dmitrytrager) [inline app/helpers/permissions_helper.rb]: If there are several types of permissionable, we could have several presenters and move this logic there. Same here and in `permissions_check_box_value` method.

File: app/presenters/users/approvals_presenter.rb [hunk 4]
```diff
@@ -26,6 +26,33 @@ def gon
...
 
+    def input_options
+      if apps_page?
+        { placeholder: "App name", class: "validate devs_typeahead", required: true }
+      else
+        { placeholder: "email@email.com", class: "validate", required: true }
+      end
+    end
+
+    def pending_friends
```
Comment (dmitrytrager) [inline app/presenters/users/approvals_presenter.rb]: Does this method do something similar to [to_sentence](https://apidock.com/rails/Array/to_sentence) ?

File: app/presenters/users/friends_presenter.rb [hunk 5]
```diff
@@ -53,6 +53,10 @@ def form_range_filter(text, from)
         device_id: device, from: from, to: Time.zone.today), method: :get)
     end
 
+    def friend_name
+      friend.username.present? ? friend.username : friend.email.split("@").first
```
Comment (dmitrytrager) [inline app/presenters/users/friends_presenter.rb]: Possible to delegate this to friend

File: spec/presenters/users/dashboards_presenter_spec.rb [hunk 6]
```diff
@@ -166,4 +167,25 @@
       expect(user).to have_received(:public_info_hash)
     end
   end
+
+  describe "visited_countries_title" do
+    context "0 countries" do
+      it "returns 'No countries visited'" do
+        expect(dashboard.visited_countries_title).to eq "No countries visited"
+      end
+    end
```
Comment (dmitrytrager) [inline spec/presenters/users/dashboards_presenter_spec.rb]: could you please leave one empty line between contexts?

File: app/controllers/users/approvals_controller.rb [hunk 7]
```diff
@@ -3,10 +3,9 @@ class Users::ApprovalsController < ApplicationController
   before_action :correct_url_user?
 
   def new
+    @presenter = Users::ApprovalsPresenter.new(current_user, params[:approvable_type])
```
Comment (dmitrytrager) [inline app/controllers/users/approvals_controller.rb]: I'd suggest to give presenters more valuable names. So we could know what entity we are dealing with.

File: app/controllers/users/checkins_controller.rb [hunk 8]
```diff
@@ -12,9 +12,9 @@ def new
   end
 
   def index
-    @presenter = ::Users::CheckinsPresenter.new(current_user, params, "index")
-    @device = @presenter.device
-    render json: @presenter.json
+    @checkins_presenter = ::Users::CheckinsPresenter.new(current_user, params)
+    @device = @checkins_presenter.device
```
Comment (dmitrytrager) [inline app/controllers/users/checkins_controller.rb]: why do we need this?

File: app/controllers/users/devices_controller.rb [hunk 9]
```diff
@@ -23,14 +25,14 @@ def new
...
+    @devices_shared_presenter = ::Users::Devices::DevicesSharedPresenter.new(current_user, params)
+    gon.push(@devices_shared_presenter.shared_gon)
   end
 
   def info
-    presenter = ::Users::DevicesPresenter.new(current_user, params, 'info')
-    @device = presenter.device
-    @config = presenter.config
+    @devices_info_presenter = ::Users::Devices::DevicesInfoPresenter.new(current_user, params)
+    @device = @devices_info_presenter.device
```
Comment (dmitrytrager) [inline app/controllers/users/devices_controller.rb]: there is no much sense in presenter if you have to instantiate other variables in controller

File: app/helpers/permissions_helper.rb [hunk 10]
```diff
@@ -12,33 +12,24 @@ def permissions_device_title(device)
...
   end
 
-  def permissions_control_class(permissionable)
-    if permissionable.class != Permission
-      " master-switches permissionable-id-#{permissionable.id}"
-    else
-      " normal-switches permissionable-id-#{permissionable.id}"
-    end
+  def permissions_switch_class(control_object)
+    control_object.class != Permission ? "master" : "permission-switch"
```
Comment (dmitrytrager) [inline app/helpers/permissions_helper.rb]: maybe it's better to check "==" instead?

File: app/presenters/users/dashboards_presenter.rb [hunk 11]
```diff
@@ -1,22 +1,15 @@
...
+    end
 
-      # These are public
-      @most_frequent_areas = fogged_city_count.first(5)
-      @percent_change = @checkins.percentage_increase('week')
-      @weeks_checkins_count = weeks_checkins.count
-      @most_used = most_used_device
-      @last_countries_loaded = last_countries
+    def most_frequent_areas
+      checkins.hash_group_and_count_by(:fogged_city).first(5)
```
Comment (dmitrytrager) [inline app/presenters/users/dashboards_presenter.rb]: move "5" to constant, please

File: app/presenters/users/dashboards_presenter.rb [hunk 12]
```diff
@@ -26,7 +19,7 @@ def most_used_device
     def last_countries
       last_countries_sql = "created_at IN(SELECT MAX(created_at) FROM checkins INNER JOIN devices ON"\
         " checkins.device_id = devices.id WHERE devices.user_id = #{@user.id} GROUP BY country_code)"
-      @checkins.where(last_countries_sql).first 10
+      checkins.where(last_countries_sql).first 10
```
Comment (dmitrytrager) [inline app/presenters/users/dashboards_presenter.rb]: the same for "10"

File: app/presenters/users/dashboards_presenter.rb [hunk 13]
```diff
@@ -38,37 +31,48 @@ def gon
...
       end
     end
 
-    def weeks_checkins
-      @checkins.where(created_at: 1.week.ago..Time.now)
-    end
-
     def months_checkins
-      @checkins.where(created_at: 1.month.ago..Time.now).limit(200).sample(100)
+      checkins.where(created_at: 1.month.ago..Time.current).limit(200).sample(100)
```
Comment (dmitrytrager) [inline app/presenters/users/dashboards_presenter.rb]: and other magic numbers

File: app/presenters/users/devices/devices_index_presenter.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    def initialize(user, params)
+      @user = user
+      @params = params
+      @devices = index_devices
+    end
+
+    def index_devices
+      user_devices.geocode_last_checkins
+      device_ids = user_devices.last_checkins.map { |checkin| checkin["device_id"] }
+      devices = user_devices.index_by(&:id).values_at(*device_ids)
```
Comment (dmitrytrager) [inline app/presenters/users/devices/devices_index_presenter.rb]: Does it mean that we need only devices from last checkins?

File: app/presenters/users/devices/devices_shared_presenter.rb [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        device: device.public_info,
+        user: device.user.public_info_hash,
+        checkin: gon_shared_checkin
+      }
+    end
+
+    private
+
+    def checkin
+      @checkin ||= device.checkins.before(device.delayed.to_i.minutes.ago).limit(1)
```
Comment (dmitrytrager) [inline app/presenters/users/devices/devices_shared_presenter.rb]: good expression to put into method

File: app/presenters/users/devices/devices_shared_presenter.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    end
+
+    def gon_shared_checkin
+      return unless checkin.present?
+      if device.fogged?
+        checkin.select("id", "created_at", "updated_at", "device_id", "fogged_lat AS lat", "fogged_lng AS lng",
```
Comment (dmitrytrager) [inline app/presenters/users/devices/devices_shared_presenter.rb]: I'd suggest to move field set into constant

File: app/presenters/users/devices/devices_show_presenter.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    attr_reader :checkins
+    attr_reader :filename
+    attr_reader :date_range
+
+    def initialize(user, params)
+      @user = user
+      @params = params
+      @device = Device.find(params[:id])
+      @date_range = checkins_date_range
+      return unless (download_format = params[:download])
```
Comment (dmitrytrager) [inline app/presenters/users/devices/devices_show_presenter.rb]: typo here? should it be "==" ?

File: app/views/users/devices/show.html.erb [hunk 18]
```diff
@@ -3,35 +3,35 @@
...
       </a>
-      <%= link_to("Delete device", user_device_path(current_user.url_id, @presenter.device), { method: :delete, class: "btn red white-text", data: { confirm: 'Are you sure?' } })%>
+      <%= link_to("Delete device", user_device_path(current_user.url_id, @device_show_presenter.device), { method: :delete, class: "btn red white-text", data: { confirm: 'Are you sure?' } })%>
       <a id="checkinNow" class="btn green white-text" href="#" data-turbolinks="false">Checkin now</a>
       <a id="checkinFoggedNow" class="btn green white-text" href="#" data-turbolinks="false">Checkin now (fogged)</a>
     </div>
   </div>
 
-  <div id="delete<%=@presenter.device.id%>" class="modal delete-modal">
+  <div id="delete<%=@device_show_presenter.device.id%>" class="modal delete-modal">
```
Comment (dmitrytrager) [inline app/views/users/devices/show.html.erb]: please, always leave spaces after `<%=` and before `%>`

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -234,6 +234,17 @@
...
       end
     end
+
+    context "display_name" do
+      it "converts a user's email if their username is empty" do
+        user.update(username: "")
+        expect(user.display_name).to eq user.email.split("@").first
+      end
+
+      it "return username if present" do
```
Comment (dmitrytrager) [inline spec/models/user_spec.rb]: it returns

File: spec/presenters/users/devices/devices_info_presenter_spec.rb [hunk 20]
```diff
@@ -0,0 +1,51 @@
+require "rails_helper"
+
+describe ::Users::Devices::DevicesInfoPresenter do
+  subject(:info_presenter) { described_class.new(user, id: device.id) }
+  let(:user) { FactoryGirl.create(:user) }
+  let(:device) do
+    d = FactoryGirl.create(:device, user_id: user.id)
```
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_info_presenter_spec.rb]: How about
```ruby
FactoryGirl.create(:device, user_id: user.id) do |obj|
  obj.config = FactoryGirl.create(:config)
end
```
?

File: spec/presenters/users/devices/devices_show_presenter_spec.rb [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    it "returns checkins converted to param provided" do
+      checkins
+      show_presenter = described_class.new(user, id: device.id, download: "gpx")
+      expect(show_presenter.checkins).to eq device.checkins.to_gpx
+    end
+  end
+
+  describe "filename" do
+    it "returns nil if no download params" do
+    end
```
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: leave empty strings between specs, please

File: spec/presenters/users/devices/devices_show_presenter_spec.rb [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+    it "returns checkins converted to param provided" do
+      checkins
+      show_presenter = described_class.new(user, id: device.id, download: "gpx")
+      expect(show_presenter.checkins).to eq device.checkins.to_gpx
+    end
+  end
+
+  describe "filename" do
+    it "returns nil if no download params" do
```
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: do we need empty specs?

File: spec/presenters/users/devices/devices_show_presenter_spec.rb [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+      expect(show_presenter.device).to be_kind_of Device
+    end
+  end
+
+  describe "checkins" do
+    it "returns nil if no download params" do
+    end
+
+    it "returns checkins converted to param provided" do
+      checkins
```
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: I'd suggest to use `let!` instead

File: spec/presenters/users/devices/devices_show_presenter_spec.rb [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+  end
+
+  describe "device" do
+    it "returns a device" do
+      expect(show_presenter.device).to be_kind_of Device
+    end
+  end
+
+  describe "checkins" do
+    it "returns nil if no download params" do
```
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: ?

File: spec/presenters/users/devices/devices_show_presenter_spec.rb (hunk 23 above)
Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: Here and everywhere you need to set some data

File: app/presenters/users/devices/devices_index_presenter.rb (hunk 14 above)
Comment (t-harps) [inline app/presenters/users/devices/devices_index_presenter.rb]: We want all devices to be shown but we want devices to be in ordered by last checkin. So the device that you used most recently (i.e. with the most recent checkin) is shown first.

File: spec/presenters/users/devices/devices_show_presenter_spec.rb (hunk 23 above)
Comment (t-harps) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: Even though I only need 'checkins' in two tests? seems like overkill to use let! and have them loaded for every test. Or do you mean put the let! inside the context where the checkins are needed?

Comment (dmitrytrager) [inline spec/presenters/users/devices/devices_show_presenter_spec.rb]: If it's only 2 tests, then yes, second option.

File: app/presenters/users/devices/devices_index_presenter.rb (hunk 14 above)
Comment (dmitrytrager) [inline app/presenters/users/devices/devices_index_presenter.rb]: Thanks

File: app/presenters/users/devices/devices_shared_presenter.rb (hunk 15 above)
Comment (t-harps) [inline app/presenters/users/devices/devices_shared_presenter.rb]: Do you mean this is ok as is or it should be extracted out? We call ```device.checkins.before(device.delayed.to_i.minutes.ago)``` in one other place so should I move this logic to the model?

Comment (dmitrytrager) [inline app/presenters/users/devices/devices_shared_presenter.rb]: I mean I'd suggest to extract. You call device twice here. It's already a good reason.

**Pull request merged**